### DISCUSSION TITLE
Trouble with Update Row in automations

### DISCUSSION THREAD

**Discussion by rhrish** [Budibase/budibase#3001] [Help] [answered]
I've been having trouble with update a row automation. I use MariaDB for my application.

So here is how I've setup my automation.

- A cron task that runs every 10 minutes.
- A Query Rows task that queries a table, with a filter criteria looking for records that have 'job_done' field as 'No'. It is limited to one record.
- A filter condition that checks if {{ length steps.1.rows }} > 0 to make sure we have a match.
- An email automation that uses the fields from the row to send out an email.
- An Update Row task that changes the 'job_done' field to 'Yes' if the email was sent successfully and also logs the JSONstringified response from the email task into a 'log' field.

Here are the problems that I'd faced.
- Adding new automation tasks always gets added on top - I saw that this has already been raised as a bug. As a workaround I've been adding all the task steps in reverse order so I get the desired order.

- For anything related to handlebars, when I want to compare a boolean, what is the value I use? Do I use true/false or True/False? If you've noticed I've setup the condition as {{ length steps.1.rows }} 'Greater Than' 0 instead of {{ steps.1.rows }} 'Equals' true. I am not sure if the latter works properly - as far as I've used it, I haven't got correct results.

- Is there a way to add multiple email IDs, CCs and BCCs in this step?

**Update Row automation troubles** - Update row somehow didn't work and the output JSON showed that the record could not be found/"404 not found"

- For this particular table, the _rev field didn't show up as an input in the automation interface.

- The rowId field showed up in the interface and I supplied the value {{ steps.1.rows.0._id }} as input. However, the response I kept getting on this step was 404 Not found in the Output JSON. In the same Output JSON, the value of docId seemed to have a double url_encoded version of the rowID. 
- For example,  the input JSON value of {{ steps.1.rows.0._id }} shows '%5B1%5D'. 
- The docId in output JSON shows up as '%255B1%255D'.
- I tried giving the input as {{ url_decode steps.1.rows.0._id }} as the rowId but this didn't work as well. Overall, my update row was not happening.

- When a MariaDB table has a TEXT type field, it doesn't show up in the Update row automation form. In my case, I had about eight fields in the table - 2 of them are TEXT fields. When I changed their type to VARCHAR, they showed up on the automation form.

I have managed to resolve some of these issues by moving a lot of code to Stored Procedures/Triggers in MariaDB but please do let me know if there are other temporary workarounds for any of these problems. 

Thanks!

**Comment by rhrish** [comment]
If it helps, here is the structure of the table I used.

EmailQueueTB

id INT PRIMARY KEY,
full_name VARCHAR(255),
email VARCHAR(255),
books_and_authors TEXT,
price DECIMAL(7,2),
job_done VARCHAR(3) NOT NULL DEFAULT 'No',
log_text TEXT,
created_at TIMESTAMP,
updated_at DATETIME

**Comment by mjashanks** [comment, marked as answer]
Hi @rhrish 

I have successfully recreated your automation, albeit using internal tables and not SQL

> Adding new automation tasks always gets added on top

This is fixed in the latest release, which has just been released - so go ahead and update

> when comparing a boolean...

just use `true`. However in your case, `{{ length steps.1.rows }} greater than 0` is correct, as `step.1.rows` is a number, not a `boolean`. I tried this and it worked for me.

> Is there a way to add multiple emails

nope - but that's a good shout. we should have `cc` and `bcc`, and allow email fields to be comma separated.

> or this particular table, the _rev 

I don't have `_rev` either, but I don't seem to need it - are you saying that something is requiring it. I am on the latest version, maybe that's the difference here.

>  the input JSON value of {{ steps.1.rows.0._id }} shows '%5B1%5D'.

Yes, this is a known issue that we are planning on resolving next month. In the meantime, can you use your actual primary key column, instead of `_id`? e.g. do you have a `id` or `uuid` or `key` column?

> When a MariaDB table has a TEXT type field...

Sounds like a bug - I will investigate and log it

**Reply by mjashanks** [reply]
PS - a little annoyance: https://github.com/Budibase/budibase/issues/3005

Bug with filters at the moment - when you are working with them sometimes it causes the screen to freeze. Refresh the page and it's all ok - will get this fixed.

**Comment by rhrish** [comment]
Hi @mjashanks 

Thanks for all your responses. 

> Yes, this is a known issue that we are planning on resolving next month. In the meantime, can you use your actual primary key column, instead of _id? e.g. do you have a id or uuid or key column?

I did try it with ID field as well and it didn't work - which is why I went on to try with _id. Anyhow I have just updated my application to 0.9.166. I am gonna try out a few things and get back.

Thanks again for this amazing platform!

**Comment by rhrish** [comment]
Hello @mjashanks

Here's an update on the row update automation.

I recreated the entire automation again and still keep encountering the same problem.

This is the output of the Row Update automation.

```
{
  "success": false,
  "response": {
    "error": "not_found",
    "reason": "missing",
    "status": 404,
    "name": "not_found",
    "message": "missing",
    "docId": "5"
  }
}
```

I tried the following variations of row ID.

```
{{ steps.1.rows.0.id }}
{{ steps.1.rows.0._id }}
{{ url_decode steps.1.rows.0.id }}
```
and for all of these variations of inputs, I keep getting the same 'not_found' response in the output.

Is there something else I might be missing?

**Reply by mjashanks** [reply]
Hey @rhrish - I can reproduce this on Postgres. I have logged it here: https://github.com/Budibase/budibase/issues/3090

Hopefully we can get this fixed asap.
